### PR TITLE
Fixes #16294 - Remove noVNC from vendors

### PR DESCRIPTION
**Pull request by UXabre**
Things to consider while reviewing:
 - By default, the background of an RFB instance is a fashionable shade of gray; however, when my other PR would be merged, this background will be automatically removed and thus look like discussed/approvded... however
 - There's a consideration to make (probably by the UX team): by default, the new RFB instance automatically clips to the containing borders, meaning a scrollbar is shown inside the canvas area, as soon as the remote desktop would bleed out. Currently, the scrollbar is "placed" on the content; which also moves the statusbar around. My preference goes the scrollbar inside the canvast, because it leaves the rest of the page untouched. In that case, I'd make a different class for VNC UI, which only removes the background and adds the backdrop shadow to the canvas instead. + there are scaling modes available as well, so, again, might be good to involve the UX team to consider the new possibilities?

Tested with oVirt as VNC provider, with two sources (VMs):
 - CentOS 7.5
 - Windows 2012R2

Seems to be working fine on my end

### CODE REVIEW

File: webpack/assets/javascripts/bundle_novnc.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    const port = $('#vnc').attr('data-port');
+    const password = $('#vnc').attr('data-password');
+ 
+    let url;
+    if ($('#vnc').data('encrypt') === 'true') {
+      url = 'wss';
+    } else {
+      url = 'ws';
+    }
+    url += '://' + host + ':' + port;
```
[the remark this answers is not in the document]
Comment (UXabre) [reply webpack/assets/javascripts/bundle_novnc.js]: Does this not break compatibility with IE11? (noVNC works in IE11)
Unless we don't want to keep supporting IE11?

[the remark this answers is not in the document]
Comment (ekohl) [reply webpack/assets/javascripts/bundle_novnc.js]: I think that webpack will run it through babel that should make it work but I'll refer to @theforeman/ui-ux to confirm/deny this.

[the remark this answers is not in the document]
Comment (UXabre) [reply webpack/assets/javascripts/bundle_novnc.js]: Oh, that would be great, I'll check it out from my side :-)

[the remark this answers is not in the document]
Comment (UXabre) [reply webpack/assets/javascripts/bundle_novnc.js]: Seems to work in IE11 using template strings, will push later today :-)

**Comment by UXabre** [comment]
A first glance, the failed tests don't seem to be related to my change here. 
Also, fixed a bug in my initial commit which prevented the console from being loaded correctly + resolved all style issues as well :-)

File: webpack/assets/javascripts/bundle_novnc.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+function disconnectFinished() {
+  showStatus('failed', 'Disconnected');
+}
+
+function connectFinished() {
+  showStatus('normal', 'Connected');
+}
+
+$(document).on('ContentLoad', () => {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/bundle_novnc.js:47]: this will now execute on every page, you should probably exit this function early once there is no vnc in the page?

File: webpack/assets/javascripts/bundle_novnc.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+  if (vncScreen.length) {
+    $('#sendCtrlAltDelButton').on('click', sendCtrlAltDel);
+
+    const host = window.location.hostname;
+    const port = $('#vnc').attr('data-port');
+    const password = $('#vnc').attr('data-password');
+
+    let url;
+    if ($('#vnc').data('encrypt') === 'true') {
```
Comment (ekohl) [inline webpack/assets/javascripts/bundle_novnc.js]: How about:
```javascript
const protocol = $('#vnc').data('encrypt') === 'true' ? 'wss' : 'ws';
const host = window.location.hostname;
const port = $('#vnc').attr('data-port');
const url = `${protocol}://${host}:${port}`;
```

File: webpack/assets/javascripts/bundle_novnc.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  rfb.sendCtrlAltDel();
+  return false;
+}
+
+function showStatus(state, message) {
+  let level;
+  const status = $('#noVNC_status');
+  const ctrlAltDeleteButton = $('#ctrlAltDelButton');
+
+  switch (state) {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/bundle_novnc.js]: probably make sense to extract to another function or use some sort of CONST?

File: webpack/assets/javascripts/bundle_novnc.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  if (typeof message !== 'undefined') {
+    status.attr('class', `col-md-12 label label-${level}`);
+    status.html(message);
+  }
+}
+
+function securityFailed(e) {
+  let msg = '';
+  if ('reason' in e.detail) {
+    msg = `New connection has been rejected with reason: ${e.detail.reason}`;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/bundle_novnc.js]: strings are not extracted?

File: webpack/assets/javascripts/bundle_novnc.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      credentials: {
+        password,
+      },
+    });
+
+    rfb.addEventListener('connect', connectFinished);
+    rfb.addEventListener('disconnect', disconnectFinished);
+    rfb.addEventListener('securityfailure', securityFailed);
+
+    showStatus('disconnected', 'Loading...');
```
Comment (ohadlevy) [inline webpack/assets/javascripts/bundle_novnc.js]: strings should be extracted.

**Review by ohadlevy — CHANGES_REQUESTED** [review]
thanks for your contribution!

it looks like there is an javascript error of some kind, so all integration tests are failing.

this looks like a positive first step, I do believe this needs to end up being a react component for viewing the console, but would be happy to merge this work prior.

I do wonder whats the impact on bundle size, and if we should consider starting to splitting the bundle again or not, i'm OK with delaying this decision further, but imho we would end up moving to that long term.

lastly, referencing the work patternfly-react to include a [console support](https://github.com/patternfly/patternfly-react/tree/master/packages/patternfly-3/react-console)

**Comment by Rohoover** [comment]
@UXabre 

Do you mind posting a screenshot for me to see? Thank you!

File: webpack/assets/javascripts/bundle_novnc.js (hunk 4 above)
Comment (UXabre) [reply webpack/assets/javascripts/bundle_novnc.js]: Perhaps a const lookup table might do the trick here; is actually the old code that I reworked just a little bit.

File: webpack/assets/javascripts/bundle_novnc.js (hunk 2 above)
Comment (UXabre) [reply webpack/assets/javascripts/bundle_novnc.js:47]: Hi @ohadlevy , thank you for the review. Is was wondering, I actually did the check on the line 62, if there is no id noVNC_screen, I don't execute anything else. Is there a better way? Should I check on path rather than on content? The best path is probably to make this or use a react component; but I'd like to descope this for now and make a separate ticket for this as I'd like to make an RFC on the whole console experience beforehand; if that's okay for you :-)

File: webpack/assets/javascripts/bundle_novnc.js (hunk 3 above)
Comment (UXabre) [reply webpack/assets/javascripts/bundle_novnc.js]: Nice proposal, will change the code to this :-)

**Comment by UXabre** [comment]
Another question, should the additional licensing information for NoVNC still be addressed in the Readme? Otherwise I can clean this up as well while we're at it :-)

image

**Comment by ekohl** [comment]
I think that can be removed.

We can also look at removing websockify and rely on a package. It's in Debian Stretch and Ubuntu. The one in EPEL is too old but we can open an issue to update it (and ship an updated on in our repos in the mean time). That's probably a separate effort from this PR though.

**Comment by UXabre** [comment]
I've made the changes as requested :-) as promised, I'd like to further continue working on these console parts (use patternfly console component, extend supported console-types, update spice-html5 console to latest version (!), enhance it with clipboard support, clearer errors, cooler animations :p, ...)

however, more importantly, the tests still seem to fail with te (recurring) error message: "Can't find variable: tfm"; I double checked by running one of the failing integration tests on the current develop head, and the results seem to be exactly the same... I started investigating what the root cause could be, this surely takes time as I'm still learning to navigate the codebase... however I fear that's outside the scope of this PR?

**Comment by ohadlevy** [comment]
@UXabre can you try running the integration tests locally? e.g.
``` RAILS_ENV=test NODE_ENV=production rake jenkins:integration TESTOPTS=-v   ```

I myself many times uses chrome driver to do this kind of tests, maybe consider trying out the #5770 pr without the headless option. I remember @glekner did something similar recently - maybe he could share his setup?

**Comment by UXabre** [comment]
Hey @ohadlevy , I did some tests and the issue seems to be triggered by NoVNC/RFB being imported. If I remove the import, everything seems to work just fine. In the previous VNC code, the RFB module was loaded dynamically, as part of the template. In my proposed change, I bundle all VNC code and initialize the RFB component when the page contains a VNC component. 

I don't really get a clear error message what the problem seems to be, but my hunch is that this has something to do with PhantomJS; It's not very clear to me how I can validate my assumption and use Chromium as my environment? (although that code works fine in chrome, under normal usage, meaning: using the frontend)

**Comment by ohadlevy** [comment]
if you run the rake task i suggested above, it should run phatomjs for you? then you could use something like http://phantomjs.org/troubleshooting.html ?

**Comment by UXabre** [comment]
SO, finally, I've managed to debug what was going on in PhantomJS. The problem is so easy it was hard to find: PhantomJS doesn't support ES6 (which is known), however, the noVNC package uses let; which is not supported in PhantomJS. 

I fixed it by transpiling the novnc module with webpack, I'll push that change but perhaps somebody knows a better way? _IMHO: I'm not sure if there should be a conversation on the use of PhantomJS, as it's development is currently suspended, an ES6 support is still not in, if we shouldn't test using "real" browsers like firefox/chromium instead? They support headless modes and are in line with most standards (which increases the importance of https://github.com/theforeman/foreman/pull/5770)_

Thanks @ohadlevy for the help in debugging this issue! Thanks everyone for reviewing, feel like we're almost there :-)

BTW @Rohoover, I seem to have overlooked your message, sorry for that, was focused on the technical aspect, I've actually updated this commit to keep the previously agreed upon UI change for VNC consoles in here as well (so the fashionable shad of gray is no more). Idea is still to use a react component for the console, until then I don't want to change too much on the UI actually. Will make an RFC for this change with some things I already have in mind for this. Anywho, it will still look like this:
after
(this is the stable's UI so far, was done to fix an issue for VNC in which the cursor was 10px misplaced: https://github.com/theforeman/foreman/pull/5979)

**Comment by UXabre** [comment]
Is see 2 tests are still failing, but doesn't look like it has anything to do with my changes? Not sure if anyone had the time yet to review these changes?

**Review by ohadlevy — DISMISSED** [review]

**Comment by ohadlevy** [comment]
@UXabre can you try adding novnc to config/webpack.vendor.js too? thanks

**Comment by ohadlevy** [comment]
@UXabre something like (my editor suggested the other change as well)
```diff
diff --git a/config/webpack.vendor.js b/config/webpack.vendor.js
index bd71747a2..5998fef62 100644
--- a/config/webpack.vendor.js
+++ b/config/webpack.vendor.js
@@ -65,4 +65,5 @@ module.exports = [
   'jstz',
   'urijs',
   'uuid',
+  '@novnc/novnc/core/rfb',
 ];
diff --git a/webpack/assets/javascripts/bundle_novnc.js b/webpack/assets/javascripts/bundle_novnc.js
index 1d6720d89..9f67a6937 100644
--- a/webpack/assets/javascripts/bundle_novnc.js
+++ b/webpack/assets/javascripts/bundle_novnc.js
@@ -29,7 +29,10 @@ function showStatus(state, message) {
 function securityFailed(e) {
   let msg = '';
   if ('reason' in e.detail) {
-    msg = window.Jed.sprintf(__('New connection has been rejected with reason: %'), e.detail.reason);
+    msg = window.Jed.sprintf(
+      __('New connection has been rejected with reason: %'),
+      e.detail.reason,
+    );
   } else {
```

**Comment by UXabre** [comment]
Hi @ohadlevy, thanks for re-reviewing; I've added the package to config/webpack.vendor.js, however, I actually added @novnc/novnc as the package here as I thought that made more sense? Or should I only add the file here that I explicitly included in my novnc_bundle.js file?

**Comment by ohadlevy** [comment]
Does it work? It didn't work for me as you can't import that afaiu

**Comment by UXabre** [comment]
@ohadlevy no, you were right, doesn't work at all actually, changed it back to your proposal; shouldn't have doubted you :-(

**Review by ohadlevy — DISMISSED** [review]

File: package.json [hunk 7]
```diff
@@ -62,6 +62,7 @@
     "phantomjs-prebuilt": "^2.1.0"
   },
   "dependencies": {
+    "@novnc/novnc": "^1.0.0",
```
Comment (ekohl) [inline package.json:65]: https://github.com/theforeman/foreman-packaging/pull/3052 adds the package.

File: config/webpack.config.js [hunk 8]
```diff
@@ -80,7 +80,7 @@ module.exports = env => {
       rules: [
         {
           test: /\.js$/,
-          exclude: /node_modules/,
+          exclude: /node_modules(?!\/@novnc)/,
```
Comment (ekohl) [inline config/webpack.config.js:84]: As someone still mostly unfamiliar with webpack: why is this needed?

Comment (ohadlevy) [reply config/webpack.config.js:84]: I believe its because of https://github.com/theforeman/foreman/pull/6056#issuecomment-423301264

Comment (ekohl) [reply config/webpack.config.js:84]: Those are the things that deserve an inline comment.

Comment (ohadlevy) [reply config/webpack.config.js:84]: @UXabre do you mind adding a comment please?

Comment (UXabre) [reply config/webpack.config.js:84]: Sure! Will do so tonight, should've done this right away. The reason is indeed as ohadlevy mentioned.

**Comment by UXabre** [comment]
Seems "the hound" is a bit over-active today? Anything I can do to _kick_ the hound and retrigger it?

File: config/webpack.config.js [hunk 9]
```diff
@@ -80,7 +80,8 @@ module.exports = env => {
       rules: [
         {
           test: /\.js$/,
-          exclude: /node_modules/,
+          // Include novnc in webpack, transpiling is needed for phantomjs (which does not support ES6) to run tests
+          exclude: /node_modules(?!\/@novnc)/,
```
Comment (ohadlevy) [inline config/webpack.config.js:84]: @sharvit can you ack this change? thanks.

**Review by ohadlevy — APPROVED** [review]

**Comment by ohadlevy** [comment]
@ekohl ack?

**Review by ekohl — APPROVED** [review]

**Comment by ohadlevy** [comment]
thanks @UXabre, it will released in 1.20.

File: config/webpack.config.js (hunk 8 above)
Comment (sharvit) [reply config/webpack.config.js:84]: noVNC decision not to include dist files in the npm is confusing.
They are supply a script to build the dist files:
https://github.com/novnc/noVNC/blob/master/docs/LIBRARY.md#conversion-of-modules
https://github.com/novnc/noVNC/blob/master/utils/use_require.js

But I guess it's easier to build it with our webpack instead.

We should follow up with a new PR to centralize the none excluded npm packages in a nice array so we can manage them easily.
